### DISCUSSION TITLE
How to make actions to copy and paste selected text in Textarea?

### DISCUSSION THREAD

**Discussion by Tim-Cao** [Textualize/textual#4051] [Q&A] [answered]
```python
from textual import on, work
from textual.app import App, ComposeResult, events
from textual.widgets import *
from textual.binding import Binding
import pyperclip


class ScriptApp(App):
    BINDINGS = [
        Binding("ctrl+p", "copy_to_clipboard"),
        Binding("ctrl+v", "paste_from_clipboard"),
    ]

    def compose(self) -> ComposeResult:
        yield TextArea("Test Sceniaro 1:", id="textarea-1")
        yield TextArea("Test Sceniaro 1:", id="textarea-2")
        yield Output()

    @work(exclusive=True, thread=True)
    def action_copy_to_clipboard(self) -> None:
        self.query_one(Output).begin_capture_print()

        text_to_copy = self.query_one(TextArea).selected_text

        pyperclip.copy(text_to_copy)

    @work(exclusive=True, thread=True)
    def action_paste_from_clipboard(self) -> None:
        self.query_one(Output).begin_capture_print()

        self.query_one(TextArea).replace(
            pyperclip.paste(),
            self.query_one(TextArea).selection.start,
            self.query_one(TextArea).selection.end,
        )


class Output(RichLog):
    def __init__(self) -> None:
        super().__init__(highlight=True, markup=True)

    @on(events.Print)
    def on_print(self, event: events.Print) -> None:
        if event.text.strip():
            self.write(event.text)


app = ScriptApp()

if __name__ == "__main__":
    app.run()

```
Run this app, it has two Textarea widgets. The actions fail since `TooManyMatches: Call to only_one resulted in more than one matched node`

**Comment by davep** [comment]
Your question doesn't really have anything to do with `TextArea` or clipboards; this is just a bug in your code when you come to doing a query. You're composing multiple widgets of the same type (in this case two `TextArea`) into your application then doing a `query_one` on just the type alone to get one back. The problem is that `query_one` can't possibly know which *one* you want, because you don't tell it.

You'll notice you actually gave your `TextArea` their own IDs; you should use that to perform the query. For example:

```python
self.query_one("#textarea-1", TextArea)
```

**Reply by Tim-Cao** [reply]
Hi @davep, I would like to make these two functions for current TextArea widget I'm working on. There are multiple TextArea widgets in my application. I don't think query by id works for my case.
```python
    @work(exclusive=True, thread=True)
    def action_copy_to_clipboard(self) -> None:
        self.query_one(Output).begin_capture_print()

        text_to_copy = self.query_one(TextArea).selected_text

        pyperclip.copy(text_to_copy)

    @work(exclusive=True, thread=True)
    def action_paste_from_clipboard(self) -> None:
        self.query_one(Output).begin_capture_print()

        self.query_one(TextArea).replace(
            pyperclip.paste(),
            self.query_one(TextArea).selection.start,
            self.query_one(TextArea).selection.end,
        )
```

**Reply by TomJGooding** [reply]
> There are multiple TextArea widgets in my application.

As Dave has patiently tried to explain, the [query_one](https://textual.textualize.io/api/dom_node/#textual.dom.DOMNode.query_one) method gets a single (the one in query_one) widget. 

How do you expect your code to work, if your code contains not one, but multiple widgets of the same type?

**Reply by Tim-Cao** [reply]
Is it possible to query TextArea widget where the cursor is on? @TomJGooding

**Reply by Zaloog** [reply]
@Tim-Cao I guess you can check if a certain TextArea widget is focused and then use the text from that.
For more infos on attributes/properties of widgets you can check the docs. Or use textual-dev and have the console run in parallel to see what events are sent and are useful to get the Widget of your choice based on certain state/attribute

**Reply by Tim-Cao** [reply]
The ideal solution is to make ctrl+insert work. https://github.com/Textualize/textual/discussions/4050
Since the ctrl+shift+c also quit the app.

**Reply by TomJGooding** [reply, marked as answer]
I think I understand the confusion now. The `query_one` method just gets a widget, it doesn't know what widget is currently focused.

You could check if the `TextArea` is focused like @Zaloog suggested. 

But actually if this binding shouldn't do anything if a different type of widget is focused, wouldn't it make more sense to add the binding to the TextArea, rather than the App?

```python
class TextAreaWithCopyPaste(TextArea):
    BINDINGS = [
        Binding("ctrl+p", "copy_to_clipboard"),
        Binding("ctrl+v", "paste_from_clipboard"),
    ]

    def action_copy_to_clipboard(self) -> None:
        text_to_copy = self.selected_text
        # etc...
```

**Reply by davep** [reply]
I'm just going to second and emphasise what Tom is saying here: if you want bindings that act on the currently-focused widget, and especially if the bindings only make sense for certain widgets (there's no need to paste into a currently-focused `Button`, for example), then the thing to do is to extend the widget in question and add the bindings and functionality there.

**Reply by Tim-Cao** [reply]
The solution works.
```python
class TextAreaWithCopyPaste(TextArea):
    BINDINGS = [
        Binding("ctrl+p", "copy_to_clipboard"),
        Binding("ctrl+v", "paste_from_clipboard"),
    ]

    def action_copy_to_clipboard(self) -> None:
        text_to_copy = self.selected_text
        # etc...
```
Thanks @TomJGooding
